GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs

```diff
@@ -384,86 +311,99 @@ private void Update()
+                // don't update status while repository is being modified by GitExt,
+                // repository status will change after these actions.
                 UICommandsSource.UICommands.RepoChangedNotifier.IsLocked ||
                 (GitVersion.Current.RaceConditionWhenGitStatusIsUpdatingIndex && Module.IsRunningGitProcess()))
             {
-                _statusIsUpToDate = false; // tell that computed status isn't up to date
+                _statusIsUpToDate = false; // schedule new update when command is finished
                 return;
             }
 
+            _workTreeWatcher.EnableRaisingEvents = true;
             _commandIsRunning = true;
             _statusIsUpToDate = true;
             _previousUpdateTime = Environment.TickCount;
 
+            // Schedule update every 5 min, even if we don't know that anything changed
+            CalculateNextUpdateTime(PeriodicUpdateInterval);
+
+            // capture a consistent state in the main thread
+            var statusCancellationToken = _statusCancellationToken;
+            IGitModule module = Module;
+
             ThreadHelper.JoinableTaskFactory.RunAsync(
                 async () =>
                 {
                     try
                     {
-                        await TaskScheduler.Default;
-
-                        _ignoredFilesPending = _ignoredFilesAreStale;
-
-                        // git-status with ignored files when needed only
-                        var cmd = GitCommandHelpers.GetAllChangedFilesCmd(!_ignoredFilesPending, UntrackedFilesMode.Default, noLocks: true);
-                        var output = Module.RunGitCmd(cmd);
-                        var changedFiles = GitCommandHelpers.GetStatusChangedFilesFromString(Module, output);
+                        try
+                        {
+                            await TaskScheduler.Default;
+                            if (!statusCancellationToken.IsCancellationRequested)
+                            {
+                                var cmd = GitCommandHelpers.GetAllChangedFilesCmd(true, UntrackedFilesMode.Default, noLocks: true);
+                                var output = module.RunGitCmd(cmd);
+                                var changedFiles = GitCommandHelpers.GetStatusChangedFilesFromString(module, output);
 
-                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
 
-                        UpdatedStatusReceived(changedFiles);
+                                UpdatedStatusReceived(changedFiles);
+                            }
+                        }
+                        catch
+                        {
+                            try
+                            {
+                                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(statusCancellationToken);
```

Comment:
Here, a CancellationException will override the caught exception if the token is cancelled. Since we have dropped the support for  ignored files, we no longer perform an expensive job here. Hence I would give up the `statusCancellationToken` at all and replace it with a simple check that the captured module has not been changed while fetching the repo status.

```csharp

            ThreadHelper.JoinableTaskFactory.RunAsync(
                async () =>
                {
                    try
                    {
                        try
                        {
                            await TaskScheduler.Default;
                            var cmd = GitCommandHelpers.GetAllChangedFilesCmd(true, UntrackedFilesMode.Default, noLocks: true);
                            var output = module.RunGitCmd(cmd);
                            var changedFiles = GitCommandHelpers.GetStatusChangedFilesFromString(module, output);

                            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync;
                            if(!ModuleHasChanged())
                            {
                                UpdatedStatusReceived(changedFiles);
                            }
                        }
                        catch
                        {
                            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync;
                            if(!ModuleHasChanged())
                            {
                                CurrentStatus = GitStatusMonitorState.Stopped;
                            }

                            throw;
                        }
                    }
                    finally
                    {
                        _commandIsRunning = false;
                    }
                })
                .FileAndForget();

                boolean ModuleHasChanged()
                {
                    return module != Module;
                }


```

Reply:
Changed, but I would have preferred to keep this
git-status can still take many seconds. I also have some thoughts about reintroducing --ignored=matching at least as a test.

Code is simpler though.